README.md
# tcpser-dummy
tcpser with dummy functionality

The Windows 64 and 32 bit prebuilt binaries are in tcpser64.zip and tcpser32.zip respectively

Read the other README file for more info.
```
Usage: ./tcpser <parameters>
  -p   port to listen on (defaults to 6400)
  -t   trace flags: (can be combined)
    's' = modem input
    'S' = modem output
    'i' = IP input
    'I' = IP output
  -l   0 (NONE), 1 (FATAL) - 7 (DEBUG_X) (defaults to 0)
  -L   log file (defaults to stderr)

  The following can be repeated for each modem desired
  (-s, -S, and -i will apply to any subsequent device if not set again)

  -d   serial device (e.g. /dev/ttyS0). Cannot be used with -v
  -v   tcp port for VICE RS232 (e.g. 25232). Cannot be used with -d
  -s   serial port speed (defaults to 38400)
  -S   speed modem will report (defaults to -s value). Setting to 555 or speed*23 enables dummy mode
  -I   invert DCD pin
  -n   add phone entry (number=replacement)
  -a   filename to send to local side upon answer
  -A   filename to send to remote side upon answer
  -c   filename to send to local side upon connect
  -C   filename to send to remote side upon connect
  -N   filename to send when no answer
  -B   filename to send when modem(s) busy
  -T   filename to send upon inactivity timeout
  -i   modem init string (defaults to '', leave off 'at' prefix when specifying)
  -D   direct connection (follow with hostname:port for caller, : for receiver)
  -S 555 dummy mode, exits after sending CONNECT, useful for connecting Windows to serial internet
  -S speed*23 dummy mode, exits after sending CONNECT SPEED, useful for connecting Windows to serial internet
```


debian/changelog
tcpser (1.0rc12-2) unstable; urgency=low

  [ Jari Aalto ]
  * QA upload.
  * Remove deprecated dpatch and upgrade to packaging format "3.0 quilt"
    (Closes: #664359).
  * Update to Standards-Version to 3.9.4, debhelper to 9, ${misc:Depends}.
  * Add build-arch and build-indep targets; use dh_prep in rules file.
  * Fix copyright-with-old-dh-make-debian-copyright (Lintian).
  * Move Alioth repository from Bzr to Git (git-buildpackage).
  * Use hardened build flags
    http://wiki.debian.org/ReleaseGoals/SecurityHardeningBuildFlags

  [ tony mancill ]
  * Update d/watch file.  Thank you to Bart Martens.
  * Add patch to address FTBFS with "ld --as-needed". (Closes: #641411)
    - Thank you to Matthias Klose.

 -- tony mancill <tmancill@debian.org>  Fri, 03 May 2013 21:27:26 -0700

tcpser (1.0rc12-1) unstable; urgency=low

  * New upstream release
  * debian/tcpser.1: fixed typos, added -v option, added square
    brackets where necessary
  * src/init.c: reverted "Fixed typo in usage summary"
  * debian/patches/01-usage-typo.dpatch: moved patch to dpatch
  * Makefile: reverted "Added install rule to makefile"
  * debian/rules, debian/install: moved install rule here
  * debian/control: updated Standards-Version, added dpatch to
    Build-Depends, added Vcs-Bzr field, added Homepage field
  * debian/watch: added
  * debian/copyright: updated year to 2008

 -- Peter Collingbourne <pcc03@doc.ic.ac.uk>  Mon,  4 Feb 2008 12:51:33 +0000

tcpser (1.0rc10-1) unstable; urgency=low

  * Initial release (Closes: #376993)
  * Added install rule to makefile
  * Wrote man page
  * Fixed typo in usage summary

 -- Peter Collingbourne <pcc03@doc.ic.ac.uk>  Tue,  4 Jul 2006 18:17:01 -0700



src/modem_core.c
#include <unistd.h>
#include <stdlib.h>       // for atoi

#include "getcmd.h"
#include "debug.h"
#include "modem_core.h"

unsigned char* mdm_responses[37];

int mdm_init() {
  mdm_responses[MDM_RESP_OK]="OK";
  mdm_responses[MDM_RESP_RING]="RING";
  mdm_responses[MDM_RESP_ERROR]="ERROR";
  mdm_responses[MDM_RESP_CONNECT]="CONNECT";
  mdm_responses[MDM_RESP_NO_CARRIER]="NO CARRIER";
  mdm_responses[MDM_RESP_CONNECT_1200] = "CONNECT 1200";
  mdm_responses[MDM_RESP_NO_DIALTONE] = "NO DIALTONE";
  mdm_responses[MDM_RESP_BUSY] = "BUSY";
  mdm_responses[MDM_RESP_NO_ANSWER] = "NO ANSWER";
  mdm_responses[MDM_RESP_CONNECT_0600] = "CONNECT 0600";
  mdm_responses[MDM_RESP_CONNECT_2400] = "CONNECT 2400";
  mdm_responses[MDM_RESP_CONNECT_4800] = "CONNECT 4800";
  mdm_responses[MDM_RESP_CONNECT_9600] = "CONNECT 9600";
  mdm_responses[MDM_RESP_CONNECT_7200] = "CONNECT 7200";
  mdm_responses[MDM_RESP_CONNECT_12000] = "CONNECT 12000";
  mdm_responses[MDM_RESP_CONNECT_14400] = "CONNECT 14400";
  mdm_responses[MDM_RESP_CONNECT_19200] = "CONNECT 19200";
  mdm_responses[MDM_RESP_CONNECT_38400] = "CONNECT 38400";
  mdm_responses[MDM_RESP_CONNECT_57600] = "CONNECT 57600";
  mdm_responses[MDM_RESP_CONNECT_115200] = "CONNECT 115200";
  mdm_responses[MDM_RESP_CONNECT_234000]="CONNECT 230400";
  return 0;
}


int get_connect_response(int speed, int level) {
  if(level == 0) {
    return MDM_RESP_CONNECT;
  }
  switch (speed) {
    case 115200:
      return MDM_RESP_CONNECT_115200;
    case 57600:
      return MDM_RESP_CONNECT_57600;
    case 38400:
      return MDM_RESP_CONNECT_38400;
    case 19200:
      return MDM_RESP_CONNECT_19200;
    case 9600:
      return MDM_RESP_CONNECT_9600;
    case 4800:
      return MDM_RESP_CONNECT_4800;
    case 2400:
      return MDM_RESP_CONNECT_2400;
    case 1200:
      return MDM_RESP_CONNECT_1200;
    case 600:
      return MDM_RESP_CONNECT_0600;
    case 555:
      /*Dummy Speed*/
      return MDM_RESP_CONNECT;
  }
  return MDM_RESP_CONNECT;
}

void mdm_init_config(modem_config* cfg) {
  int i=0;
  cfg->send_responses=TRUE;
  cfg->connect_response=0;
  cfg->response_code_level=4;
  cfg->text_responses=TRUE;
  cfg->echo=TRUE;
  cfg->cmd_mode=TRUE;
  cfg->conn_type=MDM_CONN_NONE;
  cfg->off_hook=FALSE;
  cfg->line_ringing=FALSE;
  cfg->cur_line_idx=0;

  for(i=0;i<100;i++) {
    cfg->s[i]=0;
  }
  cfg->s[2]=43;
  cfg->s[3]=13;
  cfg->s[4]=10;
  cfg->s[5]=8;
  cfg->s[6]=2;
  cfg->s[7]=50;
  cfg->s[8]=2;
  cfg->s[9]=6;
  cfg->s[10]=14;
  cfg->s[11]=95;
  cfg->s[12]=50;

  cfg->crlf[0]=cfg->s[3];
  cfg->crlf[1]=cfg->s[4];
  cfg->crlf[2]=0;

  cfg->dial_type=0;
  cfg->last_dial_type=0;
  cfg->disconnect_delay=0;

  cfg->pre_break_delay=FALSE;
  cfg->break_len=0;

  cfg->memory_dial=FALSE;
  cfg->dsr_active=FALSE;
  cfg->dsr_on=TRUE;
  cfg->dcd_on=FALSE;
  cfg->found_a=FALSE;
  cfg->cmd_started=FALSE;
  cfg->allow_transmit=TRUE;
  cfg->invert_dsr=FALSE;
  cfg->invert_dcd=FALSE;

  cfg->config0[0]='\0';
  cfg->config1[0]='\0';

  dce_init_config(cfg);
  sh_init_config(cfg);
  line_init_config(cfg);
}

int get_new_cts_state(modem_config *cfg, int up) {
  return MDM_CL_CTS_HIGH;
}


int get_new_dsr_state(modem_config *cfg, int up) {
  if(cfg->dsr_on == TRUE)
    return (cfg->invert_dsr == TRUE?MDM_CL_DSR_LOW:MDM_CL_DSR_HIGH);
  if((up == TRUE && cfg->invert_dsr == FALSE) 
     || (up == FALSE && cfg->invert_dsr==TRUE)
    )
    return MDM_CL_DSR_HIGH;
  else
    return MDM_CL_DSR_LOW;
}


int get_new_dcd_state(modem_config *cfg, int up) {
  if(cfg->dcd_on == TRUE)
    return (cfg->invert_dcd == TRUE?MDM_CL_DCD_LOW:MDM_CL_DCD_HIGH);
  if((up == TRUE && cfg->invert_dcd == FALSE) 
     || (up == FALSE && cfg->invert_dcd==TRUE)
    )
    return MDM_CL_DCD_HIGH;
  else
    return MDM_CL_DCD_LOW;
}

int mdm_set_control_lines(modem_config *cfg) {
  int state=0;
  int up=(cfg->conn_type == MDM_CONN_NONE?FALSE:TRUE);

  state |=get_new_cts_state(cfg,up);
  state |=get_new_dsr_state(cfg,up);
  state |=get_new_dcd_state(cfg,up);

  LOG(LOG_INFO,"Control Lines: DSR:%d DCD:%d CTS:%d",
      ((state & MDM_CL_DSR_HIGH) != 0?1:0),
      ((state & MDM_CL_DCD_HIGH) != 0?1:0),
      ((state & MDM_CL_CTS_HIGH) != 0?1:0)
     );

  dce_set_control_lines(cfg,state);
  return 0;
}


void mdm_write_char(modem_config *cfg,unsigned char data) {
  unsigned char str[2];
  str[0]=data;
  mdm_write(cfg,str,1);
}

void mdm_write(modem_config *cfg,unsigned char data[], int len) {
  if(cfg->allow_transmit == TRUE) {
    dce_write(cfg,data,len);
  }
}

void mdm_send_response(int msg,modem_config *cfg) {
  unsigned char msgID[17];

  LOG(LOG_DEBUG,"Sending %s response to modem",mdm_responses[msg]);
  if(cfg->send_responses==TRUE) {
    mdm_write(cfg,cfg->crlf,2);
    if(cfg->text_responses==TRUE) {
      LOG(LOG_ALL,"Sending text response");
      mdm_write(cfg,mdm_responses[msg],strlen(mdm_responses[msg]));
    } else {
      LOG(LOG_ALL,"Sending numeric response");
      sprintf(msgID,"%d",msg);
      mdm_write(cfg,msgID,strlen(msgID));
    }
    mdm_write(cfg,cfg->crlf,2);
  }
}

int mdm_off_hook(modem_config *cfg) {

  LOG(LOG_INFO,"taking modem off hook");
  cfg->off_hook=TRUE;
  cfg->cmd_mode=FALSE;
  line_off_hook(cfg);
  return 0;
}

int mdm_answer(modem_config *cfg) {
  if(cfg->line_ringing == TRUE) {
    cfg->conn_type=MDM_CONN_INCOMING;
    mdm_off_hook(cfg);
    mdm_set_control_lines(cfg);
    mdm_print_speed(cfg);
  } else if(cfg->conn_type == MDM_CONN_INCOMING) {
    // we are connected, just go off hook.
    mdm_off_hook(cfg);
    mdm_set_control_lines(cfg);
  } else {
    mdm_disconnect(cfg);
  }
  return 0;
}

int mdm_print_speed(modem_config *cfg) {
  int speed;

  switch(cfg->connect_response) {
    case 2:
      speed=cfg->dte_speed;
      break;
    default:
      speed=cfg->dce_speed;
      break;

  }
  if (speed!=555){
    if (!(speed%23)){
      /* Dummy Mode (23 Mode) */
      mdm_send_response(get_connect_response((speed/23),cfg->response_code_level),cfg);
      printf("Dummy mode (23 mode) completed, exiting...\n");
      exit(0);
    }
    else{
      mdm_send_response(get_connect_response(speed,cfg->response_code_level),cfg);
    }
  }
  else{
      /* Dummy Mode (555 Mode) */
      mdm_send_response(get_connect_response(speed,cfg->response_code_level),cfg);
      printf("Dummy mode completed, exiting...\n");
      exit(0);
  }
  return 0;
}

int mdm_connect(modem_config* cfg) {
  mdm_off_hook(cfg);
  if(cfg->conn_type == MDM_CONN_NONE) {
    if(line_connect(cfg) == 0) {
      cfg->conn_type=MDM_CONN_OUTGOING;
      mdm_set_control_lines(cfg);
      mdm_print_speed(cfg);
    } else {
      cfg->conn_type=MDM_CONN_OUTGOING;   // so disconnect will print NO CARRIER
      mdm_disconnect(cfg);
    }
  }
  return 0;
}


int mdm_listen(modem_config *cfg) {
  return line_listen(cfg);
}


int mdm_disconnect(modem_config* cfg) {
  int type;

  LOG_ENTER();
  LOG(LOG_INFO,"Disconnecting modem");
  cfg->cmd_mode=TRUE;
  cfg->off_hook=FALSE;
  cfg->break_len=0;
  cfg->line_ringing = FALSE;
  cfg->pre_break_delay=FALSE;
  if(0 == line_disconnect(cfg)) {
    type=cfg->conn_type;
    cfg->conn_type=MDM_CONN_NONE;
    mdm_set_control_lines(cfg);
    if(type != MDM_CONN_NONE) {
      mdm_send_response(MDM_RESP_NO_CARRIER,cfg);
      usleep(cfg->disconnect_delay * 1000);
    }
    cfg->rings=0;
    mdm_listen(cfg);
  } else {
    // line still connected.
  }
  LOG_EXIT();
  return 0;
}

int mdm_parse_cmd(modem_config* cfg) {
  int done=FALSE;
  int index=0;
  int num=0;
  int start=0;
  int end=0;
  int cmd=AT_CMD_NONE;
  unsigned char* command=cfg->cur_line;
  unsigned char tmp[256];

  LOG_ENTER();

  LOG(LOG_DEBUG,"Evaluating AT%s",command);

  while(TRUE != done ) {
    if(cmd != AT_CMD_ERR) {
      cmd=getcmd(command,&index,&num, &start,&end);
          LOG(LOG_DEBUG,"Command: %c (%d), Flags: %d, index=%d, num=%d, data=%d-%d",
                 (cmd > -1?cmd & 0xff:' '),
                 cmd,
                 cmd  >> 8,
                 index,
                 num,
                 start,
                 end
                );
    }
    switch(cmd) {
      case AT_CMD_ERR:
        mdm_send_response(MDM_RESP_ERROR,cfg);
        done=TRUE;
        break;
      case AT_CMD_END:
        if(cfg->cmd_mode == TRUE)
          mdm_send_response(MDM_RESP_OK,cfg);
        done=TRUE;
        break;
      case AT_CMD_NONE:
        done=TRUE;
        break;
      case 'O':
      case 'A':
          mdm_answer(cfg);
          cmd=AT_CMD_END;
          done=TRUE;
          break;
      case 'B':   // 212A versus V.22 connection
        if(num > 1) {
          cmd=AT_CMD_ERR;
        } else {
          //cfg->connect_1200=num;
        }
        break;
      case 'D':
        if(end>start) {
          strncpy(cfg->dialno,command+start,end-start);
          cfg->dialno[end-start]='\0';
          cfg->dial_type=(unsigned char)num;
          cfg->last_dial_type=(unsigned char)num;
          strncpy(cfg->last_dialno,command+start,end-start);
          cfg->last_dialno[end-start]='\0';
          cfg->memory_dial=FALSE;
        } else if (num == 'L') {
          strncpy(cfg->dialno,cfg->last_dialno,strlen(cfg->last_dialno));
          cfg->dial_type=cfg->dial_type;
          cfg->memory_dial=TRUE;
          mdm_write(cfg,cfg->crlf,2);
          mdm_write(cfg,cfg->dialno,strlen(cfg->dialno));
        } else {
          cfg->dialno[0]=0;
          cfg->last_dialno[0]=0;
          cfg->dial_type=0;
          cfg->last_dial_type=0;
        }
        if(strlen(cfg->dialno) > 0) {
          mdm_connect(cfg);
        } else {
          mdm_off_hook(cfg);
        }
        done=TRUE;
        break;
      case 'E':   // still need to define #2
        if(num == 0)
          cfg->echo=FALSE;
        else if(num == 1)
          cfg->echo=TRUE;
        else {
          cmd=AT_CMD_ERR;
        }
        break;
      case 'H':
          if(num == 0) {
            mdm_disconnect(cfg);
          } else if(num == 1) {
            mdm_answer(cfg);
          } else
            cmd=AT_CMD_ERR;
          break;
      case 'I':   // Information.
        break;
      case 'L':   // Speaker volume
        if(num < 1 || num > 3)
          cmd=AT_CMD_ERR;
        else {
          //cfg->volume=num;
        }
        break;
      case 'M':   // speaker settings
        if(num > 3)
          cmd=AT_CMD_ERR;
        else {
          //cfg->speaker_setting=num;
        }
        break;
      case 'N':   // automode negotiate
        if(num > 1)
          cmd=AT_CMD_ERR;
        else {
          //cfg->auto_mode=num;
        }
        break;
      case 'P':   // defaut to pulse dialing
        //cfg->default_dial_type=MDM_DT_PULSE;
        break;
      case 'Q':   // still need to define #2
        if(num == 0)
          cfg->send_responses=TRUE;
        else if(num == 1)
          cfg->send_responses=FALSE;
        else if(num == 2)  // this should be yes orig/no answer.
          cfg->send_responses=TRUE;
        else {
          cmd=AT_CMD_ERR;
        }
        break;
      case 'S':
        strncpy(tmp,command+start,end-start);
        tmp[end-start]='\0';
        cfg->s[num]=atoi(tmp);
        switch(num) {
          case 3:
            cfg->crlf[0]=cfg->s[3];
            break;
          case 4:
            cfg->crlf[1]=cfg->s[4];
            break;
        }
        break;
      case AT_CMD_FLAG_QUERY | 'S':
        sprintf(tmp,"%s%3.3d",cfg->crlf,cfg->s[num]);
        mdm_write(cfg,tmp,strlen(tmp));
        break;
      case 'T':   // defaut to tone dialing
        //cfg->default_dial_type=MDM_DT_TONE;
        break;
      case 'V':   // done
        if(num == 0)
          cfg->text_responses=FALSE;
        else if(num ==1)
          cfg->text_responses=TRUE;
        else {
          cmd=AT_CMD_ERR;
        }
        break;
      case 'W':
          if(num > -1 && num < 3) 
            cfg->connect_response=num;
          else
            cmd=AT_CMD_ERR;
          break;
      case 'X':
          if(num > -1 && num < 5) 
            cfg->response_code_level=num;
          else
            cmd=AT_CMD_ERR;
          break;
      case 'Y':   // long space disconnect.
        if(num > 1)
          cmd=AT_CMD_ERR;
        else {
          //cfg->long_disconnect=num;
        }
        break;
      case 'Z':   // long space disconnect.
        if(num > 1)
          cmd=AT_CMD_ERR;
        else {
          // set config0 to cur_line and go.
        }
        break;
      case AT_CMD_FLAG_EXT + 'C':
        switch(num) {
          case 0:
            cfg->dcd_on=TRUE;
            mdm_set_control_lines(cfg);
            break;
          case 1:
            cfg->dcd_on=FALSE;
            mdm_set_control_lines(cfg);
            break;
          default:
            cmd=AT_CMD_ERR;
            break;
        }
        break;
      case AT_CMD_FLAG_EXT + 'K':
        // flow control.
        switch (num) {
          case 0:
            dce_set_flow_control(cfg,0);
            break;
          case 3:
            dce_set_flow_control(cfg,MDM_FC_RTS);
            break;
          case 4:
            dce_set_flow_control(cfg,MDM_FC_XON);
            break;
          case 5:
            dce_set_flow_control(cfg,MDM_FC_XON);
            // need to add passthrough..  Not sure how.
            break;
          case 6:
            dce_set_flow_control(cfg,MDM_FC_XON | MDM_FC_RTS);
            break;
          default:
            cmd=AT_CMD_ERR;
            break;
        }
        break;
      default:
        break;
    }
  } 
  cfg->cur_line_idx=0;
  return cmd;
}

int mdm_handle_char(modem_config* cfg, unsigned char ch) {
  if(cfg->echo == TRUE)
    mdm_write_char(cfg,ch);
  if(cfg->cmd_started == TRUE) {
    if(ch == (unsigned char)(cfg->s[5])) {
      if(cfg->cur_line_idx == 0 && cfg->echo == TRUE) {
        mdm_write_char(cfg,'T');
      } else {
        cfg->cur_line_idx--;
      }
    } else if(ch == (unsigned char)(cfg->s[3])) {
      // we have a line, process.
      cfg->cur_line[cfg->cur_line_idx] = 0;
      strncpy(cfg->last_cmd,cfg->cur_line,sizeof(cfg->last_cmd) - 1);
      mdm_parse_cmd(cfg);
      cfg->found_a=FALSE;
      cfg->cmd_started=FALSE;
    } else {
      cfg->cur_line[cfg->cur_line_idx++ % sizeof(cfg->cur_line)]=ch;
    }
  } else if(cfg->found_a == TRUE) {
    if(ch == 't' || ch == 'T') {
      cfg->cmd_started=TRUE;
      LOG(LOG_ALL,"'T' parsed in serial stream, switching to command parse mode");
    } else if(ch == '/') {
      LOG(LOG_ALL,"'/' parsed in the serial stream, replaying last command");
      cfg->cur_line_idx=strlen(cfg->last_cmd);
      strncpy(cfg->cur_line,cfg->last_cmd,cfg->cur_line_idx);
      mdm_parse_cmd(cfg);
      cfg->cmd_started=FALSE;
    } else if(ch!='a' && ch!='A') {
      cfg->found_a=FALSE;
    }

  } else if(ch == 'a' || ch == 'A') {
    LOG(LOG_ALL,"'A' parsed in serial stream");
    cfg->found_a=TRUE;
  } 
  return 0;
}

int mdm_clear_break(modem_config* cfg) {
  cfg->break_len=0;
  cfg->pre_break_delay=FALSE;
  return 0;
}

int mdm_parse_data(modem_config* cfg,unsigned char* data, int len) {
  int i;

  if(cfg->cmd_mode==TRUE) {
    for(i=0;i<len;i++) {
      mdm_handle_char(cfg,data[i]);
    }
  } else {
    line_write(cfg,data,len);
    if(cfg->pre_break_delay == TRUE) {
      for(i=0;i<len;i++) {
        if(data[i] == (unsigned char)cfg->s[2]) {
          LOG(LOG_DEBUG,"Break character received");
          cfg->break_len++;
          if(cfg->break_len > 3) {  // more than 3, considered invalid
            cfg->pre_break_delay=FALSE;
            cfg->break_len=0;
          }
        } else {
          LOG(LOG_ALL,"Found non-break character, cancelling break");
          // chars past +++
          mdm_clear_break(cfg);
        }
      }
    }
  }
  return 0;
}

int mdm_handle_timeout(modem_config* cfg) {
  if(cfg->pre_break_delay == TRUE && cfg->break_len == 3) {
    // pre and post break.
    LOG(LOG_INFO,"Break condition detected");
    cfg->cmd_mode=TRUE;
    mdm_send_response(MDM_RESP_OK,cfg);
    mdm_clear_break(cfg);
  } else if(cfg->pre_break_delay == FALSE) {
    // pre break wait over.
    LOG(LOG_DEBUG,"Initial Break Delay detected");
    cfg->pre_break_delay=TRUE;
  } else if(cfg->pre_break_delay == TRUE 
            && cfg->break_len > 0
           ) {
    LOG(LOG_ALL,"Inter-break-char delay time exceeded");
    mdm_clear_break(cfg);
  } else if(cfg->s[30] != 0) {
    // timeout...
    LOG(LOG_INFO,"DTE communication inactivity timeout");
    mdm_disconnect(cfg);
  }
  return 0;
}

int mdm_send_ring(modem_config *cfg) {
  LOG(LOG_DEBUG,"Sending 'RING' to modem");
  cfg->line_ringing = TRUE;
  mdm_send_response(MDM_RESP_RING,cfg);
  cfg->rings++;
  LOG(LOG_ALL,"Sent #%d ring",cfg->rings);
  if(cfg->cmd_mode == FALSE || (cfg->s[0] != 0 &&cfg->rings>=cfg->s[0])) {
    mdm_answer(cfg);
  }
  return 0;
}




src/init.c
#include <stdio.h>
#include <stdlib.h>       // for exit,atoi
#include <unistd.h>
#include "debug.h"
#include "phone_book.h"
#include "init.h"

void print_help(unsigned char* name) {
  fprintf(stderr, "Usage: %s <parameters>\n",name);
  fprintf(stderr, "  -p   port to listen on (defaults to 6400)\n");
  fprintf(stderr, "  -t   trace flags: (can be combined)\n");
  fprintf(stderr, "       's' = modem input\n");
  fprintf(stderr, "       'S' = modem output\n");
  fprintf(stderr, "       'i' = IP input\n");
  fprintf(stderr, "       'I' = IP output\n");
  fprintf(stderr, "  -l   0 (NONE), 1 (FATAL) - 7 (DEBUG_X) (defaults to 0)\n");
  fprintf(stderr, "  -L   log file (defaults to stderr)\n");
  fprintf(stderr, "\n");
  fprintf(stderr, "  The following can be repeated for each modem desired\n");
  fprintf(stderr, "  (-s, -S, and -i will apply to any subsequent device if not set again)\n");
  fprintf(stderr, "\n");
  fprintf(stderr, "  -d   serial device (e.g. /dev/ttyS0). Cannot be used with -v\n");
  fprintf(stderr, "  -v   tcp port for VICE RS232 (e.g. 25232). Cannot be used with -d\n");
  fprintf(stderr, "  -s   serial port speed (defaults to 38400)\n");
  fprintf(stderr, "  -S   speed modem will report (defaults to -s value). Setting to 555 or speed*23 enables dummy mode\n");
  fprintf(stderr, "  -I   invert DCD pin\n");
  fprintf(stderr, "  -n   add phone entry (number=replacement)\n");
  fprintf(stderr, "  -a   filename to send to local side upon answer\n");
  fprintf(stderr, "  -A   filename to send to remote side upon answer\n");
  fprintf(stderr, "  -c   filename to send to local side upon connect\n");
  fprintf(stderr, "  -C   filename to send to remote side upon connect\n");
  fprintf(stderr, "  -N   filename to send when no answer\n");
  fprintf(stderr, "  -B   filename to send when modem(s) busy\n");
  fprintf(stderr, "  -T   filename to send upon inactivity timeout\n");
  fprintf(stderr, "  -i   modem init string (defaults to '', leave off 'at' prefix when specifying)\n");
  fprintf(stderr, "  -D   direct connection (follow with hostname:port for caller, : for receiver)\n");
  fprintf(stderr, "  -S 555 dummy mode, exits after sending CONNECT, useful for connecting Windows to serial internet\n");
  fprintf(stderr, "  -S speed*23 dummy mode, exits after sending CONNECT SPEED, useful for connecting Windows to serial internet\n");
  exit(1);
}

int init(int argc, 
         char** argv, 
         modem_config cfg[],
         int max_modem, 
         int* port,
         unsigned char* all_busy,
         int all_busy_len
         ) {
  int i=0;
  int j=0;
  int opt=0;
  int trace_flags=0;
  unsigned char* tok;
  int dce_set=FALSE;
  int tty_set=FALSE;

  LOG_ENTER();
  *port=6400;
  mdm_init_config(&cfg[0]);
  cfg[0].dte_speed=38400;
  cfg[0].dce_speed=38400;

  while(opt>-1 && i < max_modem) {
    opt=getopt(argc,argv,"p:s:S:d:v:hw:i:Il:L:t:n:a:A:c:C:N:B:T:D:");
    switch(opt) {
      case 't':
        trace_flags=log_get_trace_flags();
        for(j=0;j<strlen(optarg);j++) {
          switch(optarg[j]) {
            case 's':
              trace_flags|=TRACE_MODEM_IN;
              break;
            case 'S':
              trace_flags|=TRACE_MODEM_OUT;
              break;
            case 'i':
              trace_flags|=TRACE_IP_IN;
              break;
            case 'I':
              trace_flags|=TRACE_IP_OUT;
              break;
          }
          log_set_trace_flags(trace_flags);
        }
        break;
      case 'a':
        strncpy(cfg[i].data.local_answer,optarg,sizeof(cfg[i].data.local_answer));
        break;
      case 'A':
        strncpy(cfg[i].data.remote_answer,optarg,sizeof(cfg[i].data.remote_answer));
        break;
      case 'c':
        strncpy(cfg[i].data.local_connect,optarg,sizeof(cfg[i].data.local_connect));
        break;
      case 'C':
        strncpy(cfg[i].data.remote_connect,optarg,sizeof(cfg[i].data.remote_connect));
        break;
      case 'B':
        strncpy(all_busy,optarg,all_busy_len);
        break;
      case 'N':
        strncpy(cfg[i].data.no_answer,optarg,sizeof(cfg[i].data.no_answer));
        break;
      case 'T':
        strncpy(cfg[i].data.inactive,optarg,sizeof(cfg[i].data.inactive));
        break;
      case 'i':
        strncpy(cfg[i].config0,optarg,255);
        break;
      case 'I':
        cfg[i].invert_dcd=TRUE;
        break;
      case 'p':
        *port = (atoi(optarg));
        break;
      case 'n':
        tok=strtok(optarg,"=");
        pb_add(tok,strtok(NULL,"="));
        break;
      case 'l':
        log_set_level(atoi(optarg));
        break;
      case 'L':
        log_set_file(fopen(optarg,"w+"));
        // should check to see if an error occurred...
        break;
      case 's':
        cfg[i].dte_speed=atoi(optarg);
        LOG(LOG_ALL,"Setting DTE speed to %d",cfg[i].dte_speed);
        if(dce_set==FALSE)
          cfg[i].dce_speed=cfg[i].dte_speed;
        break;
      case '?':
      case 'h':
        print_help(argv[0]);
        break;
      case 'd':
      case 'v':
        if (tty_set) {
          if (++i < max_modem) {
            dce_set=FALSE;
            mdm_init_config(&cfg[i]);
            cfg[i].dte_speed=cfg[i-1].dte_speed;
            cfg[i].dce_speed=cfg[i-1].dce_speed;
            cfg[i].dce_data.is_ip232=FALSE;
            strncpy(cfg[i].config0,cfg[i-1].config0,sizeof(cfg[i].config0));
            strncpy(cfg[i].data.local_connect,cfg[i-1].data.local_connect,sizeof(cfg[i].data.local_connect));
            strncpy(cfg[i].data.remote_connect,cfg[i-1].data.remote_connect,sizeof(cfg[i].data.remote_connect));
            strncpy(cfg[i].data.local_answer,cfg[i-1].data.local_answer,sizeof(cfg[i].data.local_answer));
            strncpy(cfg[i].data.remote_answer,cfg[i-1].data.remote_answer,sizeof(cfg[i].data.remote_answer));
            strncpy(cfg[i].data.no_answer,cfg[i-1].data.no_answer,sizeof(cfg[i].data.no_answer));
            strncpy(cfg[i].data.inactive,cfg[i-1].data.inactive,sizeof(cfg[i].data.inactive));
          } else {
            LOG(LOG_WARN,"Maximum modems defined - ignoring extra");
            break;
          }
        }
        strncpy(cfg[i].dce_data.tty,optarg,sizeof(cfg[i].dce_data.tty));
        cfg[i].dce_data.is_ip232=('v'==opt);
        tty_set=TRUE;
        break;
      case 'S':
        cfg[i].dce_speed=atoi(optarg);
        dce_set=TRUE;
        break;
      case 'D':
        cfg[i].data.direct_conn=TRUE;
        strncpy(cfg[i].data.direct_conn_num,optarg,sizeof(cfg[i].data.direct_conn_num));
        break;
    }
  }

  if (tty_set) {
    if (i < max_modem)
      ++i;
  } else {
    // no modems defined
    LOG(LOG_FATAL,"No modems defined");
    print_help(argv[0]);
  }

  LOG(LOG_DEBUG,"Read configuration for %i serial port(s)",i);

  LOG_EXIT();
  return i;
}



